src/passes/AbstractTypeRefining.cpp

```diff
@@ -106,6 +106,12 @@ struct AbstractTypeRefining : public Pass {
       }
     }
 
+    // Assume all public types are created, which makes them non-abstract and
+    // hence ignored below.
+    for (auto type : ModuleUtils::getPublicHeapTypes(*module)) {
+      createdTypes.insert(type);
+    }
```

Comment:
Do we need to make this assumption? If we assumed the public types are not allocated outside the module due to closed-world, we could still optimize the casts in the test case without changing the public types. It is probably safer to never make assumptions about public types, though 🤔

Reply:
Yeah, we could do that, if we checked and found we'd need no changes to public types. I'll add a TODO, but it seems nontrivial.